### PR TITLE
[WIP] Channels last refactor

### PR DESCRIPTION
**Pull request by jjsjann123**
Channels Last support in nvfuser

Background:
To support channels last in nvfuser with optimal performance, we want to allow dimension collapsing in generated code on channels-last tensors, which greatly simplifies indexing.
Current API in codegen only allows dimensional collapsing on neighboring axes. The unfortunate thing is that memory format design in PyTorch is implicitly marked by strides, while the semantics meaning of axes remain unchanged. i.e. A 4d tensor with axes [N, C, H, W] would have the same shape in both format, while contiguous tensor carries strides [C*H*W, H*W, W, 1] and channels-last tensor [H*W*C, 1, W*C, C].

Approach:
We identify input tensor in channels-last format and permute them to NHWC. This creates an inconsistency between codegen tensor and TorchScript tensor. Our parser handles and propagates memory format accordingly. I.e., consumes and produces channels-last inputs when it can, while transposes inputs to original format and output non-permuted outputs.
Fusion inputs/outputs in channels-last format is marked and permuted before/after fusion execution to ensure correctness on the interfacing between nvfuser and TorchScript.

- [x] add simple cpp test to ensure simplified indexing in generated code.
~add python tests to verify nhwc fp16 inputs is handled properly.~ It has been handled in recent bfloat PR

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 1]
```diff
@@ -294,10 +136,51 @@ InputsIdLookup::IdLookupReturn InputsIdLookup::lookupId(
...
+// For details  on Part_2, refer to implementation Note [ Format Bookkeeping and
+// Propagation in Parser ]
 std::vector<at::Tensor> FusionExecutorCache::runFusionWithInputs(
-    const at::ArrayRef<IValue>& inputs) {
+    const at::ArrayRef<IValue>& inputs_) {
   FUSER_PERF_SCOPE("FusionExecutorCache::runFusionWithInputs");
 
+  // permute input tensor for kernel execution. See Part_1 in Note [ Channels
+  // Last support in nvfuser ]
+  at::ArrayRef<IValue> inputs = inputs_;
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: Nit: maybe the `inputs_` parameters should be left as `inputs` and the local variable be changed to `perm_inputs` so as not to confuse the input parameter as a data member?

**Comment by jjsjann123** [comment]
bump threads for review :pray:

**Comment by naoyam** [comment]
> Current API in codegen only allows dimensional collapsing on neighboring axes.

I think we can change this. Can you please file an issue?

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 2]
```diff
@@ -294,10 +136,51 @@ InputsIdLookup::IdLookupReturn InputsIdLookup::lookupId(
...
+// memory format design in PyTorch is implicitly marked by strides, while the
+// semantics meaning of axes remain unchanged. i.e. A 4d tensor with axes [N, C,
+// H, W] would have the same shape in both format, while contiguous tensor
+// carries strides [C*H*W, H*W, W, 1] and channels-last tensor [H*W*C, 1, W*C,
+// C].
+//
+// Approach:
+// Part_1. To allow axes collapsing for channels-last format in codegen, we can
+// permute input tensor to have axes in decending order by their strides, so
+// they would be viewed as `contiguous` in codegen, hence collapsed to simple
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:155]: > we can permute input tensor to have axes in decending order by their strides,

This makes sense. Do we need to make it specific to channels last? This seems like a workaround that could be generally useful for tensors whose axes are not ordered in the descending way.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:155]: That's true. Arbitrarily permuted tensor could be handled in similar ways and it would be beneficial to support them especially for memory-format-agnostic operations like pointwise ops. And for ops like batch_norm, where semantics of axes matters, we could just permute the tensor to supported memory-format.

I'm a little bit reluctant to do that right now as: 1. there's not a real use-case that those arbitrarily permuted tensor is relevant; 2. pytorch API only natively/actively supports contiguous & channels_last.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:155]: Sounds reasonable as a start. We could make it more generic as needed.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 3]
```diff
@@ -270,17 +486,18 @@ class IrParser {
...
-            auto alpha = value_map[node->inputs()[2]->unique()];
 
-            if (alpha->isOneInt()) {
-              auto out = binaryOp(op_mapping[node->kind()].first, lhs, rhs);
-              value_map.emplace(node->output()->unique(), out);
-            } else {
-              auto out = op_mapping[node->kind()].second(lhs, rhs, alpha);
-              value_map.emplace(node->output()->unique(), out);
-            }
+            auto [format, lhs, rhs] = getConsistentValues(
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Just wanted to quickly confirm if I'm understanding correctly. In these multi-input ops if `c10::nullopt` is passed as format, the output will be c_last if any one of the input is c_last, would that be correct?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 4]
```diff
@@ -49,7 +50,199 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
+    if (!is_tensor_view_) {
+      return std::get<1>(getEntry());
+    }
+    auto [format_s, value_s] = getEntry();
+    auto val = convertValue(format, format_s, value_s);
+    vals_[format] = val;
+    return val;
+  }
+
+  // TODO: delete this and update accessor for value_map(_)
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp:146]: nit: I agree that deleting this constructor would be nicer.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 5]
```diff
@@ -49,7 +50,199 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
+class ValueHolder {
+ public:
+  // checks if given Val in target format exists.
+  bool hasValue(MemoryFormat format) const {
+    return vals_.count(format) != 0;
+  }
+
+  // returns Val in target format.
+  CgValue value(MemoryFormat format) const {
+    TORCH_INTERNAL_ASSERT(
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp:111]: nit: re-computed hash.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 6]
```diff
@@ -49,7 +50,199 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
+        std::unordered_map<int, int> permutation_axes;
+        for (const auto i : c10::irange(n_dim - 2)) {
+          permutation_axes[1 + i] = 2 + i;
+        }
+        permutation_axes[n_dim - 1] = 1; // {{1, 2}, {2, 3}, ... {n-1, 1}}
+        value_d = transpose(tv, permutation_axes);
+        break;
+      }
+      default:
+        TORCH_INTERNAL_ASSERT(false);
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: nit: error message.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 7]
```diff
@@ -49,7 +50,199 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
+    auto n_dim = tv->nDims();
+    switch (switch_pair(format_d, format_s)) {
+      case switch_pair(MemoryFormat::ChannelsLast, MemoryFormat::Contiguous): {
+        std::unordered_map<int, int> permutation_axes;
+        for (const auto i : c10::irange(n_dim - 2)) {
+          permutation_axes[n_dim - 1 - i] = n_dim - 2 - i;
+        }
+        permutation_axes[1] =
+            n_dim - 1; // {{n-1, n-2}, {n-2, n-3}, ... {1, n-1}}
+        value_d = transpose(tv, permutation_axes);
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp:203]: Please let me know if segmenter has any problem with these IO transpose ops. I don't think it will have correctness issue as is but could need updating.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 3 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Also maybe I missed this part but would we be able to handle implicit broadcast on channel dimension in this case ? e.g. 
```
 A[N,C,H,W] + B[H,W]
```

**Review by shmsong — COMMENTED** [review]
Overall looks good to me. Had some question on memory format promotion rule.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 7 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:203]: That was a thing I struggled earlier. But per Christian's suggestion, we now move the permutation logic onto the vanilla `Fusion`, which is not segmented. So we should be good there.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 3 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Good catch!!! I don't think it works, I need to update python test to cover that.

As a quick WAR, we could be disabling format propagation for implicit broadcast and permute all inputs to contiguous.
Alternatively, we can unsqueeze contiguous tensors, so they would have a broadcast dim towards the end... I need to go over the list of ops to double check if there's any exceptions.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: >Just wanted to quickly confirm if I'm understanding correctly. In these multi-input ops if c10::nullopt is passed as format, the output will be c_last if any one of the input is c_last, would that be correct?

And yes this is correct

File: torch/csrc/jit/codegen/cuda/fusion.h [hunk 8]
```diff
@@ -229,6 +229,28 @@ class TORCH_CUDA_CU_API Fusion final {
   std::unordered_set<int> getOutputAliasIndices() const;
   std::vector<std::pair<int, int>> getInputAliasIndices() const;
 
+  // mark input at index to be in channels last format
+  void setChannelsLastInputIndices(int index) {
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/fusion.h]: I talked to you about this live.  The naming here seems confusing to me where you are marking an Input as Channels last but the mechanism by which you do so is recording the index of the specific input.  Perhaps it would be less confusing if you just named the method `setChannelsLastInput()` and remove the fact that you are specifying the input via an index from the name.

**Review by kevinstephano — COMMENTED** [review]
I added a couple of comments.  The rest looked okay to me.

**Comment by jjsjann123** [comment]
Updated per review comments. bumping to get a second review/stamp :bow:

**Review by kevinstephano — APPROVED** [review]
Adding my approval.
